spyder/api/plugins.py

```diff
@@ -476,34 +767,11 @@ def register_plugin(self):
         you override this method, please don't forget to make that call
         here too.
         """
+        # FIXME: We should do this on behalf of the users on the main window
+        # and avoid "silly errors". By doing this we can also move this to
+        # the mixin and unify the information. No nneed to mention docks.
```

Comment:
I don't know how to do that, but if you know, please do it.

Reply:
Here the idea is the following:

Widgets should define actions following a defined and more structured method. This can be achieved by having a mixin that all spyder plugins must inherit to provide some facilities to make this.

A widget registers all of its actions (that we want to expose).

Then from the plugin we need to have a method to register shortcuts and actions in a single call and also have a way of registering shortcuts for actions that a widget defines. Since this is currently manual or inoinstent among different widgets, is why I suggest having a mixin to provide some facilities for widgets. 

In this process widgets should never ever use CONF, and any property os behavior that they need to define needs to also follow some standard, either by providing an options dictionary on `__init__`or providing a `setup_XXXX` like some widgets do (code editor to put an example). But again this should follow convention.

Reply:
Is `SpyderPluginWidget` a plugin or a widget or both, or do you use plugin and widget interchangeably?

How can widgets react to changes in `CONF` if you provide it as an option dictionary on `__init__`?

Reply:
>Is SpyderPluginWidget a plugin or a widget or both, or do you use plugin and widget interchangeably?

Well, it is ... both. It is a SpyderPlugin as it defines all the methods of a plugin. When docked it uses the dockwidget attribute but when undocked the plugin itself is the widget of the main window.

>How can widgets react to changes in CONF if you provide it as an option dictionary on __init__?

---

Widgets in general should communicate internal changes to options via signals

`sig_option_changed` so that it can be hook to the plguin (the thing that is aware of the 

And a plugin also should do the same so that thing can bubble down to widgets:

```python
class Plugin
    sig_config_changed   

     def __init__(self):
          self.widgeta = ChildWidgetA(self)
          self.widgetb = ChildWidgetB(self)

          self.sig_config_changed.connect(self.update_options)
          self.widgeta.sig_option_changed.connect(self.update_configuration)
          self.widgetb.sig_option_changed.connect(self.update_configuration)

     def update_options(self, options):
           """This updates options on widgets."""
           for option, value in options.items():
               self.widgeta.update_option(option, value)
               self.widgeta.update_option(option, value)

     def update_configuration(self, options):
           """This updates config with optionsfrom widgets."""
           for option, value in options.items():
               self.config.set(option, value)

class ChildWidgetA:   
    sig_option_changed

class ChildWidgetB:
    sig_option_changed
```

This the general idea that all widgets should follow. To do this I suggest some mixin for ALL spyder widgets so that we provide some basic template (the methods and machinery to update options and inner state, this could simplify some widgets a los as it would remove all the attribute definition) and provide the basic signals.

Thoughts @jitseniesen ?

Reply:
Sounds okay. To be honest, I don't remember what I was thinking. I think part of my confusion is that I am not thinking of a widget as the class that should provide an action. For me, a widget is an UI element that should be as dumb as possible. However, Spyder is not written that way (probably because such a pure approach is not practical).

When you say "widgets should never ever use CONF", I guess this does not apply to SpyderPluginWidget even though it is a widget.